<details><summary>Error Log on Linux</summary>

```
=================================== FAILURES ===================================
_________ TestSetLocaleContextManager.test_set_locale_context_manager __________
self = <test.unit.b2http.test_b2http.TestSetLocaleContextManager testMethod=test_set_locale_context_manager>
    def test_set_locale_context_manager(self):
        test_locale = locale.normalize(
            'C.utf8'
        )  # C.UTF-8 on Ubuntu 18.04 Bionic, C.utf8 on Ubuntu 22.04 Jammy
        other_locale = 'C'
    
        saved = locale.setlocale(locale.LC_ALL)
        if saved == test_locale:
            test_locale, other_locale = other_locale, test_locale
    
        locale.setlocale(locale.LC_ALL, other_locale)
>       with setlocale(test_locale):
test/unit/b2http/test_b2http.py:319: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/nix/store/pkj7cgmz66assy7l18zc7j992npb41nx-python3-3.10.12/lib/python3.10/contextlib.py:135: in __enter__
    return next(self.gen)
b2sdk/b2http.py:69: in setlocale
    yield locale.setlocale(locale.LC_ALL, name)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
category = 6, locale = 'en_US.UTF-8'
    def setlocale(category, locale=None):
    
        """ Set the locale for the given category.  The locale can be
            a string, an iterable of two strings (language code and encoding),
            or None.
    
            Iterables are converted to strings using the locale aliasing
            engine.  Locale strings are passed directly to the C lib.
    
            category may be given as one of the LC_* values.
    
        """
        if locale and not isinstance(locale, _builtin_str):
            # convert to string
            locale = normalize(_build_localename(locale))
>       return _setlocale(category, locale)
E       locale.Error: unsupported locale setting
/nix/store/pkj7cgmz66assy7l18zc7j992npb41nx-python3-3.10.12/lib/python3.10/locale.py:620: Error
```
</details>

--- a/pkgs/development/python-modules/b2sdk/default.nix
+++ b/pkgs/development/python-modules/b2sdk/default.nix
@@ -1,7 +1,8 @@
 { lib
-, arrow
+, stdenv
 , buildPythonPackage
 , fetchPypi
+, glibcLocales
 , importlib-metadata
 , logfury
 , pyfakefs
@@ -12,18 +13,19 @@
 , requests
 , setuptools-scm
 , tqdm
+, typing-extensions
 }:
 
 buildPythonPackage rec {
   pname = "b2sdk";
-  version = "1.19.0";
+  version = "1.24.0";
   format = "setuptools";
 
   disabled = pythonOlder "3.7";
 
   src = fetchPypi {
     inherit pname version;
-    hash = "sha256-aJpSt+dXjw4S33dBiMkaR6wxzwLru+jseuPKFj2R36Y=";
+    hash = "sha256-6zSjCt+J6530f1GMc/omP1zXKQKU1SDLLvslMWoqMcU=";
   };
 
   nativeBuildInputs = [
@@ -31,12 +33,13 @@ buildPythonPackage rec {
   ];
 
   propagatedBuildInputs = [
-    arrow
     logfury
     requests
     tqdm
   ] ++ lib.optionals (pythonOlder "3.8") [
     importlib-metadata
+  ] ++ lib.optionals (pythonOlder "3.12") [
+    typing-extensions
   ];
 
   nativeCheckInputs = [
@@ -44,13 +47,13 @@ buildPythonPackage rec {
     pytest-lazy-fixture
     pytest-mock
     pyfakefs
+  ] ++ lib.optionals stdenv.isLinux [
+    glibcLocales
   ];
 
   postPatch = ''
     substituteInPlace setup.py \
       --replace 'setuptools_scm<6.0' 'setuptools_scm'
-    substituteInPlace requirements.txt \
-      --replace 'arrow>=0.8.0,<1.0.0' 'arrow'
   '';
 
   disabledTestPaths = [

--- a/pkgs/development/tools/backblaze-b2/default.nix
+++ b/pkgs/development/tools/backblaze-b2/default.nix
@@ -1,13 +1,14 @@
-{ lib, python3Packages, fetchPypi }:
+{ lib, python3Packages, fetchPypi, installShellFiles }:
 
 python3Packages.buildPythonApplication rec {
   pname = "backblaze-b2";
-  version = "3.7.0";
+  version = "3.9.0";
+  format = "setuptools";
 
   src = fetchPypi {
     inherit version;
     pname = "b2";
-    sha256 = "sha256-sW6gaZWUh3WX+0+qHRlQ4gZzKU4bL8ePPNKWo9rdF84=";
+    hash = "sha256-Z9LQapWl0zblcAyMOfKhn5/O1H6+tmgiPQfAB241jqU=";
   };
 
   postPatch = ''
@@ -19,22 +20,27 @@ python3Packages.buildPythonApplication rec {
       --replace 'setuptools_scm<6.0' 'setuptools_scm'
   '';
 
-  nativeBuildInputs = with python3Packages; [
-    setuptools-scm
+  nativeBuildInputs = [
+    installShellFiles
+    python3Packages.setuptools-scm
   ];
 
   propagatedBuildInputs = with python3Packages; [
+    argcomplete
+    arrow
     b2sdk
     phx-class-registry
     setuptools
     docutils
     rst2ansi
     tabulate
+    tqdm
   ];
 
   nativeCheckInputs = with python3Packages; [
     backoff
     more-itertools
+    pexpect
     pytestCheckHook
   ];
 
@@ -46,20 +52,25 @@ python3Packages.buildPythonApplication rec {
     # require network
     "test_files_headers"
     "test_integration"
+
+    # fixed by https://github.com/Backblaze/B2_Command_Line_Tool/pull/915
+    "TestRmConsoleTool"
   ];
 
   disabledTestPaths = [
     # requires network
     "test/integration/test_b2_command_line.py"
+
+    # it's hard to make it work on nix
+    "test/integration/test_autocomplete.py"
   ];
 
   postInstall = ''
     mv "$out/bin/b2" "$out/bin/backblaze-b2"
 
-    sed 's/b2/backblaze-b2/' -i contrib/bash_completion/b2
-
-    mkdir -p "$out/share/bash-completion/completions"
-    cp contrib/bash_completion/b2 "$out/share/bash-completion/completions/backblaze-b2"
+    installShellCompletion --cmd backblaze-b2 \
+      --bash <(${python3Packages.argcomplete}/bin/register-python-argcomplete backblaze-b2) \
+      --zsh <(${python3Packages.argcomplete}/bin/register-python-argcomplete backblaze-b2)
   '';
 
   meta = with lib; {

--- a/pkgs/development/tools/backblaze-b2/default.nix
+++ b/pkgs/development/tools/backblaze-b2/default.nix
@@ -65,6 +65,7 @@ python3Packages.buildPythonApplication rec {
   meta = with lib; {
     description = "Command-line tool for accessing the Backblaze B2 storage service";
     homepage = "https://github.com/Backblaze/B2_Command_Line_Tool";
+    changelog = "https://github.com/Backblaze/B2_Command_Line_Tool/blob/v${version}/CHANGELOG.md";
     license = licenses.mit;
     maintainers = with maintainers; [ hrdinka kevincox tomhoule ];
   };

